CLIENT_SPECIFICATION.md

```diff
@@ -0,0 +1,194 @@
+
+
+## Changelog
+ - v1.0 (18th December 2018)
+	 - Initial draft
+ - v1.1 (18th December 2018)
+	 - Refined bullet points according to comments
+	 - Ensure the spec specifies the _interface_, not the workings
+ - v2.0 (12th January 2019)
+	 - Expanded notes into full sentences
+
+
+## Terminology
+In order to aid the understanding of this specification document, a number of terms will be defined in this section.
+
+### Page
+tldr-pages consists of multiple _pages_ - each of which describes a specific command.
+
+### Platform
+Pages are grouped by platform. A platform is an operating system. For example, `windows`, `linux`, `osx`. The special platform `common` contains pages for commands that work identically across all platforms.
+
+
+## Command-Line Interface
+This section describes the standardised command-line interface (CLI) for clients implementing one. Clients that do not provide a CLI can ignore this section.
+
+### Arguments
+A number of arguments MUST be supported if a CLI is implemented:
+
+Argument		        | Meaning
+------------------------|--------------------
+`--update`				| Updates the offline cache of pages.
+`--version`, `-v`		| Shows the current version of the client.
+`--list-all`, `-a`		| Lists all the pages in the current platform to the standard output. One page name per line. Additional decoration MAY be printed to the standard error.
+
+Clients MAY support additional custom arguments not documented here.
+
+Here are some examples invocations using the above flags:
+
+```bash
+tldr --update
+tldr --version
+tldr -a
+```
+
+### Page Names
+The first argument that does not start with a dash (`-`) MUST be considered the page name.
+
+In addition, page names MAY contain spaces (e.g. `git status`) - such page names MUST be transparently concatenated with dashes (`-`). For example, the page name:
+
+```
+git checkout
+```
+
+...becomes this:
+
+```
+git-checkout
+```
+
+Here are some example invocations:
+
+```bash
+tldr 7za
+tldr eyeD3
+tldr git checkout
+# In the below, "--foo" is a custom argument that takes a parameter.
+tldr --foo bar bash
+```
+
+#### Specifying the Platform
+As pages are grouped by platform, a user may want to access a platform-specific version of a page. This MUST be supported by prefixing the page name as follows:
+
+```
+windows/type
+linux/sh
+osx/brew
+```
+
+Example invocations are as follows:
+
+```bash
+tldr windows/type
+# --foo is a custom argument, as described above
+tldr --foo bar common/git merge
+tldr linux/notify-send
+```
+
+
+## Directory Structure
+This section documents the directory structure that contains the pages themselves.
+
+The master version of every page is stored inside (but not directly) the `pages` directory. Inside this directory, there is a folder for each platform - for example `windows`, `linux`, and the special `common` platform:
+
+ - `pages/`
+	 + `common/`
+	 + `linux/`
+	 + `windows/`
+	 + `osx/`
```

Comment:
should this be macos instead of osx?

Reply:
Sorry, proposal to rename osx to macos should be a separate issue. This document describes the state of the pages now, and is therefore the right thing to do.

Reply:
Probably, but it's probably worth opening an issue for that and handing that as a separate issue. The issue with that rename is that it'll break clients.

Reply:
FWIW it should probably be darwin rather than macos but I suspect that to be user _unfriendly_ 🤔 TBH this platform division is cut at wrong layer completely. Interesting bit isn't an os or kernel but userland tools family - whether it is gnu or bsd.